Verify 'upload_files' capability when displaying upload UI in media blocks

imath opened the pull request:
Closes #1536 
Closes #3672

I've been looking for a way to solve this issue #3672 and I think one possible way is to add a capability attribute to blocks and check the current user has this cap before registering the block.

## Description
In the classic editor, the upload button is disabled if the current user does not have the `upload_files` capability (eg: contributor). But when a contributor uses Gutenberg he can use "file" relative blocks. When he tries to use one of them as this role hasn't got the `upload_files` capability, the REST reply to site.url/wp-json/wp/v2/media is rest_cannot_create (403).

## How Has This Been Tested?
I have tested this PR for the contributor role. A contributor does not have the `upload_files` capability and the PR is making sure he can't use blocks involving a file upload (eg: image, core-image, audio, video, gallery).

## Screenshots (jpeg or gifs if applicable):
See #3672 for a video of the issue.

## Types of changes
I think this PR is first a bug fix, but i also think it could be interesting to be able to restrict block usage according to the user's role.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

imath commented:
I'll update the PR to fix the failing test if you think it's a good idea.

mtias commented on blocks/library/image/index.js:
```diff
@@ -59,6 +59,8 @@ registerBlockType( 'core/image', {
 		},
 	},
 
+	capability: 'upload_files',
```
Not sure we should apply the capability to the whole block in this case. It should be ok for a user to pick an existing image from the library without uploading. Seems the components themselves (media upload button, draggable upload area) are the ones that should check for the capability and change how they are displayed.

imath replied:
Thanks a lot for your review @mtias 

On one hand I totally agree a contributor should be able to pick an existing media from the WordPress Media Library.

On the other hand, he can't even do that 😢  Being logged in as a contributor, when you click on the button to pick an image, a video or an audio file, you get an empty library as the ajax action that performs the attachments query is checking for the `upload_files` capability very early. See [/wp-admin/includes/ajax-actions.php](https://core.trac.wordpress.org/browser/trunk/src/wp-admin/includes/ajax-actions.php#L2679).

So today these "media" blocks are pretty useless for contributors, that's why I suggested to remove them as it's an experience that is a bit frustrating 😬 

To only allow picking to contributors would require a WordPress core change that would be a bit more difficult than simply allowing uploading and picking just like the other roles, imho. I can think of UI/feature changes to remove the Uploader & the Upload tab of the WordPress Media Editor for instance.
Adding the `upload_files` capability to the contributor role would be much easier, but it would probably require some wide communication to WordPress users as I can imagine some are using this role for the users they wan't to restrict from uploading files...

mtias replied:
This makes sense. I think it'd be ok to remove them from the inserter, but how should we handle editing an existing post that has images? Or a post that has an image placeholder?

aduth replied:
> how should we handle editing an existing post that has images? Or a post that has an image placeholder?

@imath Is there anything which needs to be considered here?

imath replied:
@aduth This is handled into the PR actually. Images or any other media in existing post will be displayed. Just like in the classic editor the contributor can remove the image from the post content (but not from the Media Library) or edit display attributes (alignment, etc..)

imath commented on blocks/api/registration.js:
```diff
@@ -24,6 +24,17 @@ const blocks = {};
+ * @type {Object}
+ */
+const currentUser = get(
+	Object.values( pickBy( window._wpAPIDataPreload, ( value, key ) => key.indexOf( '/wp/v2/users/me' ) !== -1 ) ),
```
> You might also imagine an existing post which had been authored by another user with the capability. If we don't allow the block to be registered, it may appear as broken.

Good Point. In the classic editor this can only happen if the upper roles (author, editor, administrator), keep the post as pending. As soon as the post is published, the contributor cannot edit it anymore.

imath commented:
FWIW i've tested in the classic editor the situation where a contributor can edit a post containing an image. In this case, the contributor can edit the markup (alignment, caption, link, classes, rel attribute) and can remove the image from the content.

When it's a gallery of images, the contributor cannot see it. He sees this placeholder instead
![gallery](https://user-images.githubusercontent.com/1834524/34585834-c2222ed4-f1a1-11e7-8a6b-4bbfb261238e.png)
When he tries to edit it, he has the same issue we have here:  he can't add images to the gallery or list the images of the gallery. But he can still edit the markup and remove the gallery.
As in the existing [90 tickets about gallery](https://core.trac.wordpress.org/query?status=accepted&status=assigned&status=new&status=reopened&status=reviewing&description=~gallery&order=priority) none are talking about this "improvable" experience, it makes me think the possibility for a contributor to edit a pending post containing an image might be an edge case. 

Anyway & imho, a possible approach could be to hide the media blocks from the block pickers, so that a contributor cannot be the one who inserts one of these blocks. If another user with the `upload_files` capability added a media block and kept the post as pending. The media should be displayed to the contributor and he should be able to edit the markup, alignment or even remove the block. So this means the only change to the block there would be in the block's toolbar where the edit button shouldn't be displayed so that the contributor cannot open the Media Editor.

@aduth about WP_REST_Attachments_Controller: creating items checks for the same capability (`upload_files`). Getting media is more permissive, but do you know if it would be that permissive if it was used in the context of the Media Editor ?

imath commented:
@mtias @aduth 

In 1216849 I'm using the approach I've described in my previous [comment](#issuecomment-355416117). But instead of not showing the blocks, I'm using the feature that disables them when the property `useOnce` is `true` (eg: the more block).

So when the contributor is using the inserter he sees:

![disable](https://user-images.githubusercontent.com/1834524/34653949-a9acd2f0-f3f4-11e7-932d-5c25fc014f2f.png)

When another user with an upper role added an image or a gallery and kept the post as pending.. The contributor can see the corresponding block, remove it or modify markup, caption, link, alignment etc.. But he can't use the Edit button to open the WP Media Editor and can't transform it to another type (as the other types also require the `upload_files` capability).

![toolbars](https://user-images.githubusercontent.com/1834524/34654053-2e7b7864-f3f6-11e7-9f4e-5b557a1d990c.png)

I haven't edited the Block autocompleter because i think there's another issue to fix in `blockAutocompleter` function as it doesn't seem to check for disabled blocks. As you can see below it's possible to include more than once more block using the autocompleter.

![more-more](https://user-images.githubusercontent.com/1834524/34654075-6b87237a-f3f6-11e7-93e6-b912d3e7cdf7.png)

aduth commented:
Good stuff!

>I'm using the feature that disables them when the property useOnce is true (eg: the more block). So when the contributor is using the inserter he sees:

Do you think it will be clear to the user what's happening here? Wondering if it would be much more effort to include a tooltip explaining the disabling (which may or may not be difficult, since mouse events tend not to be fired for disabled elements).

>I haven't edited the Block autocompleter because i think there's another issue to fix in `blockAutocompleter` function as it doesn't seem to check for disabled blocks.

Perhaps related to #4097 ?

imath commented:
Thanks :)

About the tooltip, i've tried to use the Tooltip component but i'm stuck as it is not supporting [disabled elements](https://github.com/WordPress/gutenberg/blob/master/components/tooltip/index.js#L130). I think the user will understand he can't use the blocks, but i'm not sure he will figure out why, especially users who are not too aware of WordPress roles and capabilities. So I agree it's a bit too bad.

It looks like I have 2 lint errors, but i wasn't able to replicate them on my local environment, so I'll give it another look soon.

About the Autocompleter, yes i think it's best to wait for #4097 to be fixed, before adding the capability check to it.

aduth commented on editor/components/block-switcher/index.js:
```diff
@@ -17,20 +18,26 @@ import { keycodes } from '@wordpress/utils';
 		return null;
 	}
 
+	const userAllowedBlocks = allowedBlocks.filter( b => ! b.capability || userCapabilities[ b.capability ] );
```
I think there's a chance `userCapabilities` could be undefined here; likely not observed in general usage only because we [preload the `/users/me` request](https://github.com/WordPress/gutenberg/blob/75d81e9d13c065fc80d3184fe9ac58c0da0f0573/lib/client-assets.php#L792) but something which could easily break in future refactoring.

One option is to provide the [third argument to Lodash's `_.get`](https://lodash.com/docs/4.17.4#get), which is a default value in case `capabilities` is not available. We could assign this to an empty object.

Or: We could bypass the filter altogether by wrapping this in an `if` condition or including `userCapabilities` in the above condition to `return null;`.

Aside: I have a personal vendetta against abbreviated variable names. I'd rather we assign a full variable name `block` and just split this into multiple lines. The single line doesn't necessarily help readability, or at least not as much as I think the abbreviated variable name harms it.

aduth commented on editor/edit-post/modes/visual-editor/test/inserter.js:
```diff
@@ -36,8 +36,15 @@ describe( 'VisualEditorInserter', () => {
+		};
 		const wrapper = shallow(
-			<VisualEditorInserter insertBlock={ insertBlock } mostFrequentlyUsedBlocks={ mostFrequentlyUsedBlocks } />
+			<VisualEditorInserter insertBlock={ insertBlock } mostFrequentlyUsedBlocks={ mostFrequentlyUsedBlocks } user={ user } />
```
This is a long line, and I think we might consider splitting it across multiple:

```jsx
<VisualEditorInserter
	insertBlock={ insertBlock }
	mostFrequentlyUsedBlocks={ mostFrequentlyUsedBlocks }
	user={ user }
/>
```

aduth reviewed (commented):
Looping in @noisysocks , who's been working on relevant Inserter refactoring that might overlap with some of the changes proposed here (#4497).

imath commented:
Thanks for your review @aduth.

1897ae2 is applying them.

noisysocks commented on editor/components/inserter/menu.js:
```diff
@@ -95,7 +97,14 @@ export class InserterMenu extends Component {
+
+		if ( ! isDisabled && requiredCapability ) {
+			isDisabled = ! get( this.props.user.data, [ 'capabilities', requiredCapability ] );
+		}
```
Heads up: this logic has moved to `selectors.getInserterItems` (see https://github.com/WordPress/gutenberg/pull/4497).

imath commented:
Hi @noisysocks thanks for your alert about this logic move.

@aduth I just updated the PR accordingly.

noisysocks commented on editor/store/selectors.js:
```diff
@@ -1196,17 +1200,26 @@ function buildInserterItemFromReusableBlock( enabledBlockTypes, reusableBlock )
+		 * make sure he cannot use it disabling it.
+		 */
+		if ( userCapabilities && blockType.capability && ! get( userCapabilities, blockType.capability, false ) ) {
+			blockType.isDisabled = true;
```
We should be able to avoid this mutating side effect by passing `userCapabilities` into `buildInserterItemFromBlockType` and performing the check there.

imath commented:
Thanks for your feedback @noisysocks d43bbb6 should avoid the mutating side effect.

youknowriad commented:
Sorry, we kind of forgot about this PR :) Do you think you can rebase it @imath so we can do a last check before merging. I know you're probably busy with WordCamp Paris, it can probably wait for the next week.

imath commented:
Hi @youknowriad To be honest so do I! No problem, i'll rebase it this coming week-end, once the WordCamp Paris rush hours are behind us 😅

imath commented:
Hi @youknowriad it took a while longer than i thought! Sorry for the extra delay. I've just fixed merge conflicts, rebased and tested everything again 😉

gziolo commented:
It looks like that API returns an array with capabilities. Is it possible that block would require a check for more than one capability? At the moment we introduce only check for images, but I wonder if it should be represented as an array in the case when a user needs to have more permissions to edit a given block.
I'm also not convinced that the name `capability` is the best choice in here. It might be confusing for the plugin developers because it is a user's capability, not block's. It would be nice if it were clear that it is a required user's capability.

imath commented:
Hi @gziolo 

Thanks for your feedback. I can't think of an example where WordPress would require more than one capability to do be able to do something. I think using an array to check for more than one capability would introduce some complexity eg: how should values in the array be treated: the user needs to have one of listed capability or all of them ? If a developer needs more than one regular capability, he could use a custom capability  like `use_my_custom_block` and use the `map_meta_cap` filter to deal with it during the REST API request.

About the name: `userCan`, `userCapability` or something else ? I'm fine with whatever you think is best 😉

gziolo commented:
@imath what is your opinion about performing this kind of check on the server instead? We are looking into moving the block's registration to the server and having non-UI properties provided there. See related issue: https://github.com/WordPress/gutenberg/issues/2751.

imath commented:
Hi @gziolo thanks a lot for the link to the issue. For now i've just quickly give it a first look. It's very promising & interesting. I'll come back later this week-end to rebase once again the PR and i'll try to add a specific example to show what I think is needed to make this capibility check more generic and not only for the "core" media blocks. If i agree it's related, i'm not sure it's a blocker for this PR.

imath commented:
@gziolo I think the user capability check needs to be done in JavaScript because as we've seen working on this issue, it's not just about not loading the block if the user does not have the capability to `upload_files`. We need to let the block load and disable it in JavaScript. The reason is while the post is in draft mode, there's a probability a user with a higher role edits the post and adds a new media block in it, then leave it as a draft. To preserve the experience of the contributor when he comes back and edit the same draft, the inserted media block is doing his job (eg: displaying the image or the gallery) but some actions are disabled: transformation, replacement of media etc.. Just like in the classic editor the contributor can still edit the html, change the alignment or link or remove the media from the content.

That being said, I can see a real interest about doing some stuff on the server side to add this capability feature to any custom block.

User capabilities are fetched using this Rest Request: `/wp/v2/users/me?post_type=${ postType }&context=edit`

As `upload_files` is a core capability, it's fetched but let's say a custom block choose a custom capability like 'can_use_mycustom_block`, then it's probably needed to edit the `users/me` endpoint so that it performs a `current_user_can()` check on all the custom blocks custom capabilities to make sure the `map_meta_cap` filter is fired.

We had to deal with this for Multisite configs for the super admins who don't have the capabilities for a subsite (see https://github.com/WordPress/gutenberg/blob/master/lib/compat.php#L187-L232)

So having the capability attribute available into the `WP_Block_Type` would probably be be helpful to pluck all the custom capabilities from the list of registered block to check them all and eventually add them to the capabilities parameter of the `users/me` endpoint json reply.

This way, what has been done in this PR for the media blocks would work for any block.

gziolo commented:
@karmatosed proposed new UX for the media blocks in here: https://github.com/WordPress/gutenberg/issues/5456#issuecomment-376121083. Now, the question is if with custom URL for the external images, we still should disable those blocks, or rather do something the action buttons?

![output](https://user-images.githubusercontent.com/253067/37901516-58610808-30e9-11e8-9cba-a780df575965.png)

imath commented:
Awesome ! Absolutely only Upload/Media Library buttons should be disable then. Let's wait for it!

About the second part Maybe I wasn't clear or used a bad english, a contributor can still edit alignment/html or remove the media in the PR.

So far strategy was to disable the possibility to add the block, just like the core/more block (once added). But the media url field is showing it's probably not the best one.

imath commented:
@danielbachhuber Thanks for the interesting info. This pull request is a bit getting old in a bad way 😨but I can try to adapt it so that it takes this new path to deal with capabilities. I'll work on it asap.

danielbachhuber commented:
Thanks for the update @imath!

Sorry for the misdirection, but I have a slightly different direction to point you in. After I looked at your code, I came to the realization that I don't think we should introduce `wp:action-upload-files` on the Post object because it doesn't correlate with any _fields_ on the Post object (as the other `wp:action-*` metadata does).

Instead, I think we should feed `editorCanUpload` from `OPTIONS /wp/v2/media`. The presence of "POST" in the `Allows` header will give us the information we need:

<img width="1225" alt="image" src="https://user-images.githubusercontent.com/36432/40206147-acdc131e-59e3-11e8-8050-c84b5cad714a.png">

imath commented:
Got it 😃 I'll work in this direction soon. Thanks a lot for your feedback @danielbachhuber

danielbachhuber commented:
@imath How can I help get this wrapped up?

imath commented:
Hi @danielbachhuber sorry i was busy contributing to another project. Now i can't run `npm run dev` or `npm run build` anymore. I'm using a local environment with `git://develop.git.wordpress.org/`. I don't know if it's due to the fact we now need to run grunt build in WordPress. Anyway i'm getting this error again and again after many `npm install`: 

```
rimraf ./packages/*/build ./packages/*/build-module && node ./bin/packages/build.js

/path/to/gutenberg/bin/packages/get-babel-config.js:24
		...babelDefaultConfig,
		^^^
```

Unfortunately my Mac is too old to install docker :(

gziolo commented:
> Now i can't run npm run dev or npm run build anymore. I'm using a local environment with git://develop.git.wordpress.org/. I don't know if it's due to the fact we now need to run grunt build in WordPress. Anyway i'm getting this error again and again after many npm install:

@imath, this should be fixed with #7043. I bet you aren't using the latest 8.x Node, but rather some earlier version of Node 8.x :)

ntwb commented:
> this should be fixed with #7043. I bet you aren't using the latest 8.x Node, but rather some earlier version of Node 8.x

Is there an exact version known for this @gziolo ?

If known, we should bump it to that version in `package.json`, if not "core" has a minimum of `8.9.3`, bump to that maybe?

gziolo commented:
> If known, we should bump it to that version in `package.json`, if not "core" has a minimum of `8.9.3`, bump to that maybe?

I didn't check, we can bump to the version that core uses. I think that even better solution would be to expose script which is shared between Gutenberg, packages and core to make sure those versions are the same everywhere.

ntwb commented:
Sounds good @gziolo, there's currently a request to bump the minimums here:
> https://make.wordpress.org/systems/2018/05/28/upgrade-node-npm-on-build-server/

I've created an `npm-package-json-lint` issue requesting a new rule for this.

imath commented:
@gziolo & @ntwb Thanks a lot for your help !! I confirm it's all good for me. I'll update the PR asap.

imath commented:
Hi @danielbachhuber 

I've updated the PR. I think the way I get the `OPTIONS /wp/v2/media` allow header can probably be improved (eg: function names, etc). But it's working 😓 Here are some screenshots:

The Image block for a contributor

![capture d ecran 2018-06-03 a 18 02 56](https://user-images.githubusercontent.com/1834524/40888749-004cfa4c-675c-11e8-9dd5-e37658bb42f4.png)

The image block for higher roles

![capture d ecran 2018-06-03 a 18 04 47](https://user-images.githubusercontent.com/1834524/40888758-1dd1343e-675c-11e8-9d09-ea4a72740500.png)

I haven't worked on the Cover image and the Gallery (I wanted to make sure i was on the right path!), so contributors will see the following screenshot for these blocks:

![capture d ecran 2018-06-03 a 18 06 00](https://user-images.githubusercontent.com/1834524/40888768-41af36e4-675c-11e8-9bc7-066a6c3e615b.png)

Again, tests will fail! I really need to understand how to edit the snapshots.

danielbachhuber commented:
Thanks @imath !

Requesting a JavaScript review from @aduth and copy edit from `Needs Design Feedback`

aduth commented on editor/utils/editor-can-upload/index.js:
```diff
@@ -0,0 +1,48 @@
+
+	const allowMedia = getMediaAllowHeader();
+
+	return 'GET, POST' === allowMedia;
```
What's to say this isn't returned as `POST, GET` ? Or `GET,POST` ? Or `GET, POST, DELETE` ? (Now or in the future)

Should we consider these as individual verbs, where we're chiefly concerned with the presence of `POST` ?

e.g.

```js
return includes( allowMedia.split( ',' ).map( ( verb ) => verb.trim() ), 'POST' );
```

https://lodash.com/docs/4.17.10#includes

Should we just cache this boolean value, rather than the header object?

aduth commented on editor/utils/editor-can-upload/index.js:
```diff
@@ -0,0 +1,48 @@
+		.then( ( body, status, xhr ) => {
+			return xhr.getResponseHeader( 'allow' );
+		} )
+		.then( ( response ) => {
```
Does this need to be a separate `Promise#then`, or should it just be included in the one prior? If they're meant to be distinct steps, could we organize them as such with named functions?

aduth commented on core-blocks/image/edit.js:
```diff
@@ -179,31 +198,53 @@ class ImageEdit extends Component {
-							/>
-						) }
-					/>
+					{ true === editorCanUpload() && ! isExternal ? (
```
Minor: It's explicit, but I'd prefer it to just be the simpler `editorCanUpload()` truthy check. This is common enough that it stuck out to me that we might be considering some other truthy return values from which we want to distinguish `true`.

aduth commented on core-blocks/image/edit.js:
```diff
@@ -179,31 +198,53 @@ class ImageEdit extends Component {
 	render() {
 		const { attributes, setAttributes, isLargeViewport, isSelected, className, maxWidth, noticeOperations, noticeUI, toggleSelection } = this.props;
 		const { url, alt, caption, align, id, href, width, height } = attributes;
+		const isExternal = url ? parse( wpApiSettings.root ).host !== parse( url ).host : true;
```
We'd want to avoid access to globals as much as possible, or at least centralize their use. This looks a bit like a utility function which would belong in something like [the `@wordpress/url` package](https://github.com/WordPress/packages/tree/master/packages/url).

Do we want to ensure that the URL is specifically on a different host than the REST API base, or would it be enough to do something like `url.indexOf( window.location.origin ) === 0`?

aduth reviewed (changes requested):
I don't love that we're introducing a new pattern for retrieving data from the REST API, when we're already cycling through a couple others already with [`withAPIData`](https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-api-data) and [Core Data](https://github.com/WordPress/gutenberg/tree/master/core-data).

Further, the fact that we're triggering the request when the editor loads has two implications:

- The user may be presented with the wrong information, or a flickering screen if the request hasn't completed by the time they're editing an image block.
   - In fact, I don't think it'd even flicker, since this is not in any way attached to a component lifecycle, they'd be presented the wrong information and the component would never update (or would randomly update later).
- We're issuing a request whether or not the user has the intention to work with a media block.

Instead, this seems like something which could be exposed from Core Data as a selector, named e.g. `hasUploadPermissions`, where a [resolver](https://github.com/WordPress/gutenberg/blob/master/core-data/resolvers.js) for that selector would trigger the request. Then the `MediaPlaceholder` and `ImageEdit` components could be connected to select with that data ([documentation](https://github.com/WordPress/gutenberg/tree/master/packages/data#higher-order-components)), thus allowing it to update automatically and only trigger the resolver if it's present on the page.

If, at that point, we want to try to ensure the data is available as immediately as possible, we could [preload the endpoint](https://github.com/WordPress/gutenberg/blob/423dbed98af82de110af2992806dffff754c8ecb/lib/client-assets.php#L959-L966), which also has the added advantage of allowing its response to be used in more than just this one requirement.

aduth commented on editor/edit-post/modes/visual-editor/inserter.js:
```diff
@@ -45,7 +45,14 @@ export class VisualEditorInserter extends Component {
 
 	isDisabledBlock( block ) {
-		return block.useOnce && some( this.props.blocks, ( { name } ) => block.name === name );
+		const requiredCapability = get( block, 'capability' );
```
>Is `get` necessary here? (vs. `block.capability`)

I must have had an unsubmitted review previously. This code no longer exists, so the comment can be ignored.

imath commented:
@aduth Thanks a lot for your review. I've just updated the PR applying almost all your recommendations. The only thing I didn't do is adding the media `OPTIONS` API request to the preloaded endpoints because: 
- It appears using the patch it doesn't seem to be needed,
- I think preload endpoints are only `GET` requests, and i didn't dare to venture into `gutenberg_preload_api_request()` etc.. to change this.

danielbachhuber commented:
@imath It looks like there's a test failure you need to address:

<img width="891" alt="image" src="https://user-images.githubusercontent.com/36432/41476605-ec564d44-7076-11e8-9b50-4b2efcf61ce2.png">

imath commented:
Sure @danielbachhuber I will fix it asap.

imath commented:
@danielbachhuber i've just fixed merge conflicts, adapted the 4 tests and updated the 4 snapshots.

danielbachhuber reviewed (approved):
👍 I tested this locally and it works as expected. I also made a couple minor copy changes and pushed them to your branch.

danielbachhuber commented:
Given we've been awaiting on `Needs Design Feedback` for two weeks now, I'd prefer to land this as-is from a UI perspective and address any refinements at a later date.

imath commented:
Thanks a lot @danielbachhuber 👌

imath commented:
@danielbachhuber i think the tests are failing because you need to update the snapshots if you improved the strings : 

```
npm run test-unit -- --updateSnapshot --testPathPattern core-blocks/audio/test/index.js
npm run test-unit -- --updateSnapshot --testPathPattern core-blocks/video/test/index.js
npm run test-unit -- --updateSnapshot --testPathPattern core-blocks/cover-image/test/index.js
npm run test-unit -- --updateSnapshot --testPathPattern core-blocks/gallery/test/index.js
```

It should fix the failing tests.

imath commented on packages/core-data/src/resolvers.js:
```diff
@@ -78,3 +79,15 @@ export async function* getThemeSupports() {
+ * Requests Upload Permissions from the REST API.
+ */
+export async function* hasUploadPermissions() {
+	const header = await wp.apiRequest( { path: '/wp/v2/media', method: 'OPTIONS' } )
```
Thanks for your feedback, I'll refactor it asap.

imath commented:
@gziolo it's now using `@wordpress/apiRequest` package.

gziolo reviewed (dismissed):
This PR needs some more work, so moving it to 3.2 release.

> If, at that point, we want to try to ensure the data is available as immediately as possible, we could [preload the endpoint](https://github.com/WordPress/gutenberg/blob/423dbed98af82de110af2992806dffff754c8ecb/lib/client-assets.php#L959-L966), which also has the added advantage of allowing its response to be used in more than just this one requirement.

I saw the following comment:

> I think preload endpoints are only `GET` requests, and i didn't dare to venture into `gutenberg_preload_api_request()` etc.. to change this.

If we want to establish a new pattern for exposing capabilities using `OPTIONS` we need to find a way to have it preloaded or find an alternative to expose it. Maybe we can use the selector for editor settings `getEditorSettings` instead?

> Given we've been awaiting on `Needs Design Feedback` for two weeks now, I'd prefer to land this as-is from a UI perspective and address any refinements at a later date.

@karmatosed or @jasmussen, can you help?

gziolo commented on core-blocks/audio/test/index.js:
```diff
@@ -3,9 +3,20 @@
 
 describe( 'core/audio', () => {
 	test( 'block edit matches snapshot', () => {
+		// The user can upload files.
```
I don't think we really care about the state of upload buttons in terms of validating if Audio block renders properly. We only should ensure it renders properly given the default set of props. It's fine to have the edit button in the snapshot output. Can we remove this code which registers store? In my opinion, it only makes this test harder to follow.

@youknowriad mentioned that several times in different places, I just wanted to raise it again. We should seek an alternative solution for having basic validation for all blocks. Maybe e2e tests?

imath replied:
The problem I tried to resolve faking a core store is that this store is undefined during the test and I get this error https://github.com/WordPress/gutenberg/pull/4155#issuecomment-397659323

gziolo replied:
Let's skip it at the moment. I will try to tweak it myself when this lands or is about to land. It's not a blocker 👍

gziolo commented on core-blocks/cover-image/test/index.js:
```diff
@@ -3,9 +3,20 @@
 describe( 'core/cover-image', () => {
 	test( 'block edit matches snapshot', () => {
+		// The user can upload files.
+		registerStore( 'core', {
```
The same comment as above. Can we remove this code which registers store? In my opinion, it only makes this test harder to follow.

imath replied:
same reply as above 😓 It's the only way i've found to avoid the error.

gziolo commented on core-blocks/gallery/test/index.js:
```diff
@@ -3,9 +3,20 @@
 describe( 'core/gallery', () => {
 	test( 'block edit matches snapshot', () => {
+		// The user can upload files.
+		registerStore( 'core', {
```
The same comment as above. Can we remove this code which registers store? In my opinion, it only makes this test harder to follow.

imath replied:
same reply as above 😓 It's the only way i've found to avoid the error.

gziolo commented on core-blocks/image/edit.js:
```diff
@@ -176,31 +193,53 @@ class ImageEdit extends Component {
-						) }
-					/>
+					{ hasUploadPermissions && isLocal ? (
+						<MediaUpload
```
I'm wondering it this logic should be enclosed inside `<MediaUpload />` component instead. There are 6 occurrences in the code where this component is used and we are covering only 2 occurrences in this PR. It seems like all of them should have a similar check. Moving those checks to the `MediaUpload` should make it easier to maintain.

danielbachhuber replied:
@imath Do you have an opinion on ^ this question?

imath replied:
I'd say probably, but after this: https://github.com/WordPress/gutenberg/pull/4155/commits/200198f1a2e44b1da1c0f38e24396879826f3bb6 i'd also say that for the cover image the hasUploadPermissions check is before the `<toolbar>` tag and for image/video/audio it's after to fall back on a regular edit link. So i'm unsure about it 🤔

gziolo replied:
There is a lot of code duplication at the moment, because you need to releat the same if statement and fetch the same data in all those places where you could donit only once in MediaUpload component. This is a general pattern we follow, data needs should be satisfied as close to the component that depends on it as possible.

As I understand isLocal check would have to stay as it is. 

Toolbar is not an issue in here, it will simply render nothing. There should be no visual difference, we can double check it. 

The only question we need to answer is if MediaUpload can ever be rendered if user doesn’t have a proper permission. If not, then I would be in favor of having this check more linked with MediaUpload. It doesn’t need to be the same file, we can create a wrapper component similar to the pattern used in the Document sidebar, example: https://github.com/WordPress/gutenberg/blob/master/edit-post/components/sidebar/post-author/index.js.

imath replied:
Ok I will work on it.

gziolo commented on core-blocks/video/test/index.js:
```diff
@@ -3,9 +3,20 @@
 describe( 'core/video', () => {
 	test( 'block edit matches snapshot', () => {
+		// The user can upload files.
+		registerStore( 'core', {
```
The same comment as above. Can we remove this code which registers store? In my opinion, it only makes this test harder to follow.

gziolo commented on editor/components/media-placeholder/index.js:
```diff
@@ -99,22 +100,50 @@ class MediaPlaceholder extends Component {
-					onHTMLDrop={ onHTMLDrop }
-				/>
+				{ hasUploadPermissions && (
+					<div className="editor-media-placeholder__upload-controls">
```
`Placeholder` component also wraps this content with `div`:
```jsx
<div className="components-placeholder__fieldset">
	{ children }
</div>
```
What is the reason that we need another `div` wrapper?

danielbachhuber replied:
@imath ^ Do you have an answer to this question?

imath replied:
@danielbachhuber Yes, my answer is React is throwing an error otherwise. It seems you must wrap adjacent tags into a container.

gziolo replied:
You can use Fragment component instead. See other files you modified to check how it works - in general it groups items virtually but doesn’t render any HTML wrapper.

imath replied:
Thanks a lot, i'll update the PR asap.

karmatosed commented:
Regarding the design feedback not happening, please in future flag this to the team as the label was removed then re-added. I will organise a review on this today. If ever you feel something needs a review you can assign myself to a review @danielbachhuber and you can also flag it in #core-editor on chat.wordpress.org. Please let's make sure we keep consistency of the design and don't push on without reviews. Thanks for working fast here but lets try and get back a process.

jasmussen commented:
> Given we've been awaiting on Needs Design Feedback for two weeks now, I'd prefer to land this as-is from a UI perspective and address any refinements at a later date.

Sorry about that. Lots of work to do!

Here's a GIF of this branch:

![contributor](https://user-images.githubusercontent.com/1204802/41649349-aab253b6-747b-11e8-87b5-4c7e9adf03ac.gif)

This looks pretty good. Nice work. 

There's a very brief flash of the "unlimited" version of this block. That's a little jarring. Can we avoid that?

In the placeholder, can we add a little paragraph of text that says something like "As a contributor, you can only link images, you can't upload."

karmatosed commented:
Thanks @jasmussen for doing a fast review ⭐️

imath commented:
@gziolo about:
> If we want to establish a new pattern for exposing capabilities using OPTIONS we need to find a way to have it preloaded

I gave it a try in this commit: https://github.com/WordPress/gutenberg/pull/4155/commits/6a02c2bfd89083a48878f7d688222d706607cd37 and the path is now preloaded.

@jasmussen Thanks a lot for your feedback. About:
> There's a very brief flash of the "unlimited" version of this block. That's a little jarring. Can we avoid that?

I think now the path is preloaded the brief flash should have disappeared.

danielbachhuber commented:
@imath Can you resolve the merge conflict? 

It'd be great to get this landed in `3.2`. It seems like we're pretty close to the finish line.

imath commented:
Hi @jasmussen I've just added the little paragraph you requested. Well almost. I haven't mentionned the "contributor" role in the message because there can be situation where the user has a custom role. As I'm not english, the paragraph might need some improvements. Here's a screencap:

![contributor](https://user-images.githubusercontent.com/1834524/41977026-aaefe782-7a1e-11e8-9ae2-c3320109d46e.png)

FYI @gziolo I've also neutralized the edit toolbar of the cover image block for users who don't have the `upload_files` cap.

jasmussen commented:
I think that phrasing is pretty good, and it would be nice to get this in soon. Thanks for working on this, and 👍 👍 from me!

imath commented:
@danielbachhuber & @gziolo there’s now a `MediaUploadCheck` wrapper for the capability check as you recommended.

I also edited the Gallery block and the Featured image sidebar panel to use this check.

The hasUploadPermissions that is still in use in the MediaPlaceHolder is needed for the instructions messag.

aduth commented on core-blocks/cover-image/index.js:
```diff
@@ -136,21 +137,25 @@ export const settings = {
-									label={ __( 'Edit image' ) }
-									icon="edit"
-									onClick={ open }
+					{ (
```
The `{ (` is unnecessary and can be removed, allowing de-indent of the rest of the render.

aduth commented on core-blocks/image/edit.js:
```diff
@@ -219,10 +260,12 @@ class ImageEdit extends Component {
 						onError={ noticeOperations.createErrorNotice }
 						accept="image/*"
 						type="image"
+						value={ this.props.attributes }
```
Aside: It's strange we pass attributes as value.

aduth commented on core-blocks/image/edit.js:
```diff
@@ -122,6 +125,22 @@ class ImageEdit extends Component {
+		} );
+
+		if ( newUrl !== src && id ) {
+			this.props.setAttributes( {
```
We should try to avoid calling `setAttributes` multiple times, because it could incur multiple (unnecessary) renders. Instead, you could accumulate an object with which to call it once.

```js
const nextAttributes = {
	url: newUrl,
};

if ( newUrl !== src && id ) {
	Object.assign( nextAttributes, {
		alt: undefined,
		id: undefined,
		caption: undefined,
	} );
}

this.props.setAttributes( nextAttributes );
```

I see there's some precedent with `onSelectImage`. But `onSelectImage` is wrong as well 😅

aduth commented on editor/components/media-placeholder/index.js:
```diff
@@ -86,21 +88,51 @@ class MediaPlaceholder extends Component {
 			notices,
+			hasUploadPermissions,
 		} = this.props;
+		const defaultInstruction = onSelectUrl ? '' : __( 'To edit this block, you need permission to upload media.' );
```
I've noted this elsewhere in people's usage, just an FYI that if you don't want anything output, you're best to use `null`, not the empty string, as otherwise React will unnecessarily create an empty `document.createTextFragment`.

See: https://codepen.io/aduth/pen/mKeyzQ

In this case, it doesn't occur because `Placeholder` uses a truthiness check on `instructions` to determine whether to output, but a good rule of thumb all the same.

aduth commented on editor/components/media-placeholder/index.js:
```diff
@@ -86,21 +88,51 @@ class MediaPlaceholder extends Component {
-					onFilesDrop={ this.onFilesUpload }
-					onHTMLDrop={ onHTMLDrop }
-				/>
+				{ (
```
Another unnecessary `{ (`

aduth commented on editor/components/media-upload/check.js:
```diff
@@ -0,0 +1,21 @@
+		return fallback || null;
+	}
+
+	return <Fragment>{ children }</Fragment>;
```
You don't need `Fragment` here. Just `return children;`

aduth commented on editor/components/media-upload/check.js:
```diff
@@ -0,0 +1,21 @@
+	return <Fragment>{ children }</Fragment>;
+}
+
+export default compose( [
```
Minor: With just a single higher-order component, you don't strictly need `compose`. Could be:

```js
export default withSelect( ( select ) =>  {
	// ...
} )( MediaUploadCheck );
```

... As you have it makes future refactoring easier though, so balances out; no need to change.

aduth commented on editor/components/post-featured-image/index.js:
```diff
@@ -17,70 +17,80 @@ import { withSelect, withDispatch } from '@wordpress/data';
+					</MediaUploadCheck>
+				}
+				{ !! featuredImageId && media && ! media.isLoading &&
+				<MediaUploadCheck>
```
This should be indented given the condition on the previous line:

```jsx
{ !! featuredImageId && media && ! media.isLoading && (
	<MediaUploadCheck>
```

aduth commented on lib/client-assets.php:
```diff
@@ -543,28 +543,28 @@ function gutenberg_register_scripts_and_styles() {
 	}
 
-	if ( empty( $path ) ) {
+	if ( empty( $args ) || empty( $args['path'] ) || empty( $args['method'] ) ) {
```
Aside: `GET` will certainly be the majority usage here. I'd expect it to be reasonable as a default value. Maybe even to the extreme (i.e. debatable) of supporting mixed value arrays like...

```php
[
	'/get/to/path/1',
	'/get/to/path/2',
	[
		'path'   => '/options/to/path',
		'method' => 'OPTIONS',
	],
]
```

But this is less part of a public API than I originally expected, so I don't feel strongly about it at this point.

aduth commented on lib/client-assets.php:
```diff
@@ -583,7 +583,11 @@ function gutenberg_preload_api_request( $memo, $path ) {
 
-		$memo[ $path ] = array(
+		if ( 'OPTIONS' === $args['method'] ) {
+			$response = rest_send_allow_header( $response, $server, $request );
```
Is the above `rest_do_request` assignment to `$response` wasted effort for `OPTIONS` request?

imath replied:
Yes it's needed since `rest_do_request` is using `WP_REST_Server->dispatch()` and not `WP_REST_Server->serve_request()` so the needed headers for the request are never set in `dispatch` they are set in `serve_request`.

aduth commented on packages/api-request/src/middlewares/preloading.js:
```diff
@@ -36,9 +36,10 @@ const createPreloadingMiddleware = ( preloadedData ) => ( options, next ) => {
+		if ( preloadedData[ method ] && preloadedData[ method ][ path ] ) {
 			const deferred = jQuery.Deferred();
-			deferred.resolve( preloadedData[ path ].body );
+			const response = 'OPTIONS' === method ? preloadedData[ method ][ path ] : preloadedData[ method ][ path ].body;
```
We should have a unit test for this. I'm also not sure why we ought to need to special-case OPTIONS requests. Who's to say we don't want other headers from the request? Just seems like we're making inconsistency for not an entirely valid reason.

aduth commented on packages/core-data/src/actions.js:
```diff
@@ -81,3 +81,17 @@ export function receiveThemeSupportsFromIndex( index ) {
+/**
+ * Returns an action object used in signalling that Upload permissions have been received.
+ *
+ * @param  {boolean} uploadFiles Upload permissions.
```
Minor: Don't need alignment of the `@param` with `@return`.

https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/

aduth replied:
Minor: The name of this variable could better indicate it holds a boolean value, e.g. `canUploadFiles`, or even `hasUploadPermissions` since that's what you're using for the action parameter anyways. No need to be inconsistent.

aduth commented on packages/core-data/src/resolvers.js:
```diff
@@ -50,7 +51,10 @@ export async function* getEntityRecord( state, kind, name, key ) {
-	const record = await apiRequest( { path: `${ entity.baseUrl }/${ key }?context=edit` } );
+
+	// Makes sure the media will be fetched even for contributors.
+	const context = 'media' === entity.name ? 'view' : 'edit';
```
We really shouldn't have to manually assign these contexts, going to be difficult to maintain. I don't have a great solution right now; the whole context thing is causing difficulties in general.

aduth commented on packages/core-data/src/resolvers.js:
```diff
@@ -78,3 +82,19 @@ export async function* getThemeSupports() {
+ * Requests Upload Permissions from the REST API.
+ */
+export async function* hasUploadPermissions() {
+	const header = await apiRequest( { path: '/wp/v2/media', method: 'OPTIONS' } )
```
As best I can tell, the `await` here is unneeded, given how you're using `Promise#then` anyways.

talldan replied:
Without the await, I believe the value of `header` would be a promise, and then the following attempt to operate on it as a string would fail.

aduth commented on packages/core-data/src/resolvers.js:
```diff
@@ -78,3 +82,19 @@ export async function* getThemeSupports() {
+				return body.headers.Allow;
+			}
+
+			return xhr.getResponseHeader( 'allow' );
```
In what circumstance would this condition be satisfied? I cannot tell by the unit tests.

talldan replied:
@imath - I might need your help to answer this question. I'm guessing it's to satisfy particular browsers (IE)?

Is the fact that there's an xhr param to the response part of a shim we're using for fetch?

imath replied:
@tallsan I must check in details but I think what you get with preloaded path is different than what you get with a regular REST request, so I probably added this if for some reason the path is not preloaded.

imath replied:
Yes i confirm that was the reason. I did that when api-request.js, now api-fetch is used maybe it's no more needed.

imath replied:
@talldan FYI to adapt to apiFetch and be sure to get the `allow` response header if the preloaded path is not available, here's how i've adapted this part:

```patch
diff --git packages/core-data/src/resolvers.js packages/core-data/src/resolvers.js
index 104106808..7836a1aec 100644
--- packages/core-data/src/resolvers.js
+++ packages/core-data/src/resolvers.js
@@ -116,13 +116,17 @@ export async function* getEmbedPreview( state, url ) {
  * Requests Upload Permissions from the REST API.
  */
 export async function* hasUploadPermissions() {
-	const header = await apiFetch( { path: '/wp/v2/media', method: 'OPTIONS' } )
-		.then( ( body, status, xhr ) => {
-			if ( body.headers ) {
-				return body.headers.Allow;
+	const header = await apiFetch( { path: '/wp/v2/media', method: 'OPTIONS', parse: false } )
+		.then( ( response ) => {
+			if ( ! response.headers.Allow ) {
+				for ( var i of response.headers.entries() ) {
+					if ( 'allow' === i[0] ) {
+						response.headers.Allow = i[1];
+					}
+				}
 			}
 
-			return xhr.getResponseHeader( 'allow' );
+			return response.headers.Allow;
 		} );
 
 	yield receiveUploadPermissions( includes( header, 'POST' ) );
```

talldan replied:
Looks like you can use headers.get, here's what I ended up with:
```
/**
 * Requests Upload Permissions from the REST API.
 */
export async function* hasUploadPermissions() {
	const response = await apiFetch( { path: '/wp/v2/media', method: 'OPTIONS', parse: false } );

	let allowHeader;
	if ( hasIn( response, [ 'headers', 'get' ] ) ) {
		// If the request is fetched using the fetch api, the header can be
		// retrieved using the 'get' method.
		allowHeader = response.headers.get( 'allow' );
	} else {
		// If the request was preloaded server-side and is returned by the
		// preloading middleware, the header will be a simple property.
		allowHeader = get( response, [ 'headers', 'Allow' ], '' );
	}
	yield receiveUploadPermissions( includes( allowHeader, 'POST' ) );
}
```

imath replied:
Awesome work 💪👏

aduth commented on packages/core-data/src/actions.js:
```diff
@@ -81,3 +81,17 @@ export function receiveThemeSupportsFromIndex( index ) {
+ *
+ * @return {Object} Action object.
+ */
+export function receiveUdploadPermissions( uploadFiles ) {
```
Typo: `receiveUdploadPermissions` -> `receiveUploadPermissions`

aduth commented on packages/core-data/src/reducer.js:
```diff
@@ -99,6 +99,23 @@ export function themeSupports( state = {}, action ) {
+ *
+ * @return {Object} Updated state.
+ */
+export function hasUploadPermissions( state = {}, action ) {
```
I don't think an empty object is intended as the default value here. I'm assuming in current usage it will be inferred as effectively truthy (since `Boolean( {} ) === true`). Should then the default value be `true` instead? In any case, probably should be a boolean at least.

aduth commented on edit-post/hooks/components/media-upload/index.js:
```diff
@@ -192,6 +192,10 @@ class MediaUpload extends Component {
 	}
 
 	render() {
+		if ( ! this.props.render ) {
```
When / why will `this.props.render` be falsey?

imath commented:
@aduth Thanks a lot for your review. I'll be afk for the rest of July unfortunately (well it's holidays for me so: not so unfortunately), so I won't be able to work on the requested changes soon, sorry.

imath commented:
👍 Many thanks to you for your help 😍

> Le 15 août 2018 à 09:18, Daniel Richards <notifications@github.com> a écrit :
> 
> rebased against master and will start addressing the code review feedback.
> 
> Thanks @imath for your great work on this so far 👍
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

talldan commented:
https://github.com/WordPress/gutenberg/pull/4155#issuecomment-390029144

> I think we should feed editorCanUpload from OPTIONS /wp/v2/media. The presence of "POST" in the Allows header will give us the information we need:

@danielbachhuber It'd be great if you could let me know why what you've detailed here is the better way of doing things.

Specifically why is this preferred to preloading a list of the users capabilities and working with that?

imath reviewed (commented):
As this PR is pretty huge already and to avoid confusion, I think I would open another issue/PR to remove this unused preloaded path.

talldan commented:
I'm finding this PR quite hard to navigate generally with the way github collapses comments (that they collapse newer comments instead of older comments is strange).

I'm going to go through the comments and hide anything irrelevant/outdated to try to keep this focused.

I'll also squash some commits down where I can.

aduth commented:
Can this still be salvaged? Is there a desire to do so?

The PHP preloading bits, in particular, have already landed in core and will need updates there if intended to proceed.

imath commented:
@aduth Absolutely! I have a strong desire to try to improve the contributor role’s experience with Media into the next editor. 1. I’ll look into WordPress branch 5.0 and will submit a ticket there about preloading path.
2. If @talldan can carry on the work he’s been doing in #9298 & #9301 (these two PR are splitting this one in smaller parts) It would be great, otherwise I can carry on from there.
3. We’ll need to take care of new blocks :
- inline image,
- file,
- media + text

Moreover I think 4.1 is about to land with a new insert from URL popover for images/vidéo/audio blocks that is not into the media + text block, which seems a bit inconsistent to me if what I saw is confirmed.

imath committed:
Use OPTIONS Rest API Media Request to get user `upload_file` capability

Edit the preloaded path and adapt the way apiFetch is preloading the requests using the OPTIONS method.
Add a hasUploadPermissions property to core/data

imath committed:
Introduce the MediaUploadCheck component

Start using it into the MediaPlaceHolder to inform the contributor role he cannot upload media. If the block supports selecting an URL, the instruction will inform the contributor role he can only add a link to a media.

imath committed:
Display a message to the contributor about the featured image

As it is not possible to fetch the Featured image for this role, no matter the context of the REST request, display a message to inform him managing the featured image needs the upload files cap.

imath commented:
Hi @aduth 

Here's the link to the WordPress 5.0 branch ticket i've just posted about updating the preloaded paths : https://core.trac.wordpress.org/ticket/45194

I've just updated the PR so that the `upload_files` capability is checked before displaying the Upload/drag and drop etc.. controls. Here is a screenshot of how it looks like:

![capture d ecran 2018-10-25 a 22 55 45](https://user-images.githubusercontent.com/1834524/47537751-f082de00-d8c7-11e8-818a-2f3af2f6e7fb.png)

imath committed:
Enforce strict format of the preloaded path when an array is used

This is to bring into Gutenberg the improvement @danielbachhuber committed in WordPress 5.0 branch
See https://core.trac.wordpress.org/changeset/43833

talldan commented:
Thanks for picking this up again @imath - shall I close #9301 since you've now rebased this?

imath commented:
Hi @talldan yes I think we can carry on there since we have all the history (even if it’s less readable 🤗)

Obtenez Outlook pour iOS<https://aka.ms/o0ukef>

________________________________
De : Daniel Richards <notifications@github.com>
Envoyé : mercredi, octobre 31, 2018 9:18 AM
À : WordPress/gutenberg
Cc : imath; Mention
Objet : Re: [WordPress/gutenberg] Verify 'upload_files' capability when displaying upload UI in media blocks (#4155)


Thanks for picking this up again @imath<https://github.com/imath> - shall I close #9301<https://github.com/WordPress/gutenberg/pull/9301> since you've now rebased this?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/WordPress/gutenberg/pull/4155#issuecomment-434598852>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABv-HMwiJKyrZ9J8CUlusqJ3jMqi_Roqks5uqVzVgaJpZM4RLupO>.

imath committed:
Make sure the Inline Image is only available to users with the upload_files cap

youknowriad commented:
What's blocking here @talldan @imath Can we get a review?

imath commented:
@youknowriad 

Thanks for asking :) 

Now adding an image using an URL has already been merged, this PR only adds a component to make sure the user has the `upload_file` cap before outputing Upload controls. If the media block includes a control to insert an external URL, then this user can only use this control to add media (image/audio/video).

As @aduth advised preloaded paths has already been edited within WordPress branch 5.0 (see  [[43833](https://core.trac.wordpress.org/changeset/43833)] to include the `/wp/v2/media` one and to use an `OPTIONS` Request. The header of this request reply is used to set the upload permission for the user.

So to me, I only have to fix merge conflicts about `packages/format-library/src/image/index.js`

imath committed:
Update generated data-core docs

gziolo commented on packages/format-library/src/image/index.js:
```diff
@@ -46,34 +46,36 @@ export const image = {
-					/> }
-				</Fragment>
+				<MediaUploadCheck>
+					<Fragment>
```
`Fragment` component is no longer necessary here.

gziolo commented on packages/editor/src/components/media-upload/check.js:
```diff
@@ -0,0 +1,15 @@
+ */
+import { withSelect } from '@wordpress/data';
+
+export function MediaUploadCheck( { hasUploadPermissions, fallback, children } ) {
```
This line can be replaced by adding a default value to the component definition:

```diff
- export function MediaUploadCheck( { hasUploadPermissions, fallback, children } ) {
- 	const optionalFallback = fallback || null;
- 	return hasUploadPermissions ? children : optionalFallback;
+ export function MediaUploadCheck( { hasUploadPermissions, fallback = null, children } ) {
+	return hasUploadPermissions ? children : fallback;
```

gziolo commented:
I merged the latest changes added in `master` so I could retest this PR.

gziolo commented:
One of the unit tests fails after merging changes from `master`:

> ● MediaPlaceholder › renders successfully when allowedTypes property is not specified
> [1] 
> [1]     expect(jest.fn()).not.toHaveErrored(expected)
> [1] 
> [1]     Expected mock function not to be called but it was called with:
> [1]     ["Error: Uncaught [TypeError: Cannot read property 'hasUploadPermissions' of undefined]

imath commented:
@gziolo Thanks a lot for testing the PR and for your review. I'll work on applying your recommandations and fixing the unit test asap.

imath committed:
Apply @gziolo recommandations and fix a failing unit test

imath commented:
@gziolo I've fixed the unit test by making sure the core data store was not undefined.

But It seems weird to me to add a test  saying it should 'renders successfully when allowedTypes property is not specified' because it should only be the case if the onSelectURL attribute of the `<MediaPlaceholder />` component is set to true. Otherwise before rendering the upload controls you also need to check the user's upload_files capability : what this PR is precisely trying to add 😉

imath commented:
Ouch sorry but I have a very old computer, i can't install docker so i'm not able to see why a e2e test is failing.. `sysctl kern.hv_support` is returning 0.

antpb commented:
I think the tests are good now @imath . Maybe someone kicked them off again. I've seen this in some of my PRs.

youknowriad commented on packages/editor/src/components/media-placeholder/index.js:
```diff
@@ -234,4 +257,18 @@ class MediaPlaceholder extends Component {
-export default withFilters( 'editor.MediaPlaceholder' )( MediaPlaceholder );
+const applyWithSelect = withSelect( ( select ) => {
+	let hasUploadPermissions = false;
+	if ( undefined !== select( 'core' ) ) {
```
I think this check is not necessary, to be consistent with other `withSelect` usage, we should assume the "core" store to be available as it's a dependency of the editor package.

youknowriad replied:
I think I saw in the comments that it's because of the unit tests. Two suggestions:

 - Avoid running this code in unit tests (test the inner component)
 - Import the `@wordpress/core-data` in the unit tests to ensure that the store is available.

youknowriad commented on packages/editor/src/components/media-upload/check.js:
```diff
@@ -0,0 +1,19 @@
+
+export default withSelect( ( select ) => {
+	let hasUploadPermissions = false;
+	if ( undefined !== select( 'core' ) ) {
```
Same comment as above.

youknowriad reviewed (approved):
Really great work on this PR 👏 

I left some minor comments. And not blocking here but I'd appreciate an e2e test (maybe as a follow-up PR) to ensure we don't regress.

gziolo commented:
> @gziolo @aduth @youknowriad do we have a chance to finally have this fixed after almost a year working on it ? 😊

Congrats on keeping this PR updated for so long and ensuring it gets merged 💯 Thanks for this contributions. I will address comments raised by Riad in the follow-up PR.

imath commented:
Thanks a lot for your help everyone 😍🎉

youknowriad commented:
This would require some changes to be back-ported in the 5.0 branch (for the preloading changes in PHP). Up for a ticket+commit @gziolo :P

imath commented:
@youknowriad the preloading part is already in Core thanks to https://core.trac.wordpress.org/changeset/43833

youknowriad commented:
Oh good thanks @imath
